rust_src/src/multibyte.rs

```diff
@@ -42,32 +43,291 @@ use crate::{
+/// Maximum value for a single byte codepoint
+pub const MAX_1_BYTE_CHAR: u32 = 0x7F;
+
+/// Minimum value for a two byte codepoint
+pub const MIN_2_BYTE_CHAR: u32 = 0x80;
+/// Maximum value for a two byte codepoint
+pub const MAX_2_BYTE_CHAR: u32 = 0x7FF;
 
-/// Maximum character codes for several encoded lengths
-pub const MAX_1_BYTE_CHAR: Codepoint = 0x7F;
-pub const MAX_2_BYTE_CHAR: Codepoint = 0x7FF;
-pub const MAX_3_BYTE_CHAR: Codepoint = 0xFFFF;
-pub const MAX_4_BYTE_CHAR: Codepoint = 0x1F_FFFF;
-pub const MAX_5_BYTE_CHAR: Codepoint = 0x3F_FF7F;
+/// Minimum value for a three byte codepoint
+pub const MIN_3_BYTE_CHAR: u32 = 0x800;
+/// Maximum value for a three byte codepoint
+pub const MAX_3_BYTE_CHAR: u32 = 0xFFFF;
+
+/// Minimum value for a four byte codepoint
+pub const MIN_4_BYTE_CHAR: u32 = 0x10000;
+/// Maximum value for a four byte codepoint
+pub const MAX_4_BYTE_CHAR: u32 = 0x1F_FFFF;
+
+/// Minimum value for a five byte codepoint
+pub const MIN_5_BYTE_CHAR: u32 = 0x20_0000;
+/// Maximum value for a five byte codepoint
+pub const MAX_5_BYTE_CHAR: u32 = 0x3F_FF7F;
+
+pub const MIN_BYTE8_CHAR: u32 = 0x3F_FF80;
+pub const MAX_BYTE8_CHAR: u32 = 0x3F_FFFF;
 
 /// Maximum length of a single encoded codepoint
 pub const MAX_MULTIBYTE_LENGTH: usize = 5;
 
+const BYTE8_OFFSET: u32 = 0x3F_FF00;
+
+// cannot use `char`, it takes values out of its range
+#[repr(transparent)]
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
+pub struct Codepoint(u32);
+
+impl Codepoint {
+    // Equivalent to BYTE8_TO_CHAR
+    /// Create a codepoint from a raw byte. If non-ascii, byte8 encode it.
+    pub fn from_raw(byte: u8) -> Codepoint {
+        match Codepoint::from(byte) {
+            cp if cp.is_ascii() => cp,
+            cp => Codepoint::from(cp.0 + BYTE8_OFFSET),
+        }
+    }
+
+    // Equivalent to ASCII_CHAR_P
+    /// Whether the character is an ASCII value
+    pub const fn is_ascii(self) -> bool {
+        self.0 < MIN_2_BYTE_CHAR
+    }
+
+    // Equivalent to SINGLE_BYTE_CHAR_P
+    /// Whether the character is a single byte value (i.e. less than 256/0x100)
+    pub const fn is_single_byte(self) -> bool {
+        self.0 < 0x100
+    }
+
+    // Equivalent to CHAR_BYTE8_P
+    /// Whether the character is a byte8 value, i.e. an encoded raw 8bit byte.
+    pub const fn is_byte8(self) -> bool {
+        self.0 >= MIN_BYTE8_CHAR
+    }
+
+    /// The amount of bytes needed to represent the codepoint's multibyte form.
+    pub fn len_bytes(self) -> usize {
+        match self.0 {
+            0..=MAX_1_BYTE_CHAR => 1,
+            MIN_2_BYTE_CHAR..=MAX_2_BYTE_CHAR => 2,
+            MIN_3_BYTE_CHAR..=MAX_3_BYTE_CHAR => 3,
+            MIN_4_BYTE_CHAR..=MAX_4_BYTE_CHAR => 4,
+            MIN_5_BYTE_CHAR..=MAX_5_BYTE_CHAR => 5,
+            MIN_BYTE8_CHAR..=MAX_BYTE8_CHAR => 2,
+            _ => invalid_character(self),
+        }
+    }
+
+    /// Get the integer value of the codepoint
+    pub const fn val(self) -> u32 {
```

Comment:
AFAICT, `val` is only used for conversions, eg `cp.val() as c_int`. If possible, I think it would be cleaner to drop this and just implement the remaining conversions directly.

Reply:
Those conversions would be fallible though, and anything involving tryfrom would just introduce more complexity. Adding as_i32 style methods would work though, if that's what you're thinking of